File: src/main/java/org/agentic/flink/example/markets/producer/BondInventoryProducer.java
package org.agentic.flink.example.markets.producer;

import org.agentic.flink.example.markets.model.MarketRecords.Inventory;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Java twin of {@code examples-bin/markets/bond_inventory_producer.py}. Synthesizes anonymised
 * dealer bid/offer quotes and publishes JSON to the {@code fnd-inventory} Kafka topic.
 *
 * <p>Run with:
 * <pre>
 *   mvn -DskipTests package
 *   java -cp target/agentic-flink-1.0.0-SNAPSHOT.jar \
 *     org.agentic.flink.example.markets.producer.BondInventoryProducer --rate 500
 * </pre>
 */
public final class BondInventoryProducer {

  private static final List<String> DEALER_POOL = buildDealerPool();
  private static final List<String> SIDES = List.of("BID", "OFFER");
  private static final List<String> ACTIONS = List.of("UPDATE", "UPDATE", "UPDATE", "DELETE");
  private static final List<String> QUOTE_TYPES = List.of("I", "F");
  private static final List<String> MARKET_SEGMENTS = List.of("IG", "AA", "HY");
  private static final List<String> PRODUCT_CDS =
      List.of("FNDIG", "FNDAA", "FNDSC", "FNDHS", "FNDHY", "FNDPG");
  private static final int[] LEVEL_CDF = {15, 35, 70, 90, 100}; // probs 0.15/0.20/0.35/0.20/0.10 cumulative
  private static final Map<String, double[]> SEG_ANCHOR =
      Map.of("IG", new double[] {110, 8}, "AA", new double[] {112, 7}, "HY", new double[] {92, 10});
  private static final Map<String, double[]> SEG_BOUNDS =
      Map.of("IG", new double[] {85, 140}, "AA", new double[] {90, 140}, "HY", new double[] {60, 120});

  public static void main(String[] args) throws Exception {
    int rate = parseInt(args, "--rate", 500);
    int batch = parseInt(args, "--batch", 50);
    long sleepNanos = Math.max(1_000_000L, 1_000_000_000L * batch / Math.max(rate, 1));

    try (MarketProducerSupport mp =
        new MarketProducerSupport(MarketProducerSupport.defaultBootstrap(), "bond-inventory")) {
      System.out.printf(Locale.ROOT,
          "producing ~%d rows/s to fnd-inventory (batch %d)%n", rate, batch);
      while (!Thread.currentThread().isInterrupted()) {
        for (int i = 0; i < batch; i++) {
          mp.send("fnd-inventory", randomRow());
        }
        LockSupport.parkNanos(sleepNanos);
      }
    }
  }

  static Inventory randomRow() {
    ThreadLocalRandom rng = ThreadLocalRandom.current();
    String segment = MARKET_SEGMENTS.get(rng.nextInt(MARKET_SEGMENTS.size()));
    String side = SIDES.get(rng.nextInt(SIDES.size()));
    double[] anchor = SEG_ANCHOR.get(segment);
    double[] bounds = SEG_BOUNDS.get(segment);
    double nudge = "BID".equals(side) ? -0.6 : 0.6;
    double price = clamp(round3(anchor[0] + nudge + rng.nextGaussian() * anchor[1]), bounds[0], bounds[1]);
    long instrumentId = rng.nextLong(10_000_000L, 40_000_000L);
    long size = sizeBucket(rng);
    double spread = rng.nextDouble() < 0.15 ? Math.max(0, rng.nextGaussian() * 30 + 120) : 0.0;
    int level = sampleLevel(rng);
    int tier = 1 + rng.nextInt(4);
    String productCD = PRODUCT_CDS.get(rng.nextInt(PRODUCT_CDS.size()));
    String quoteType = QUOTE_TYPES.get(rng.nextInt(QUOTE_TYPES.size()));
    String action = ACTIONS.get(rng.nextInt(ACTIONS.size()));
    return new Inventory(
        DEALER_POOL.get(rng.nextInt(DEALER_POOL.size())),
        instrumentId, side, price, size, spread, level, tier, segment,
        productCD, quoteType, action, System.currentTimeMillis());
  }

  private static long sizeBucket(ThreadLocalRandom rng) {
    double r = rng.nextDouble();
    if (r < 0.25) return rng.nextInt(1, 20);
    if (r < 0.60) return rng.nextInt(20, 250);
    if (r < 0.85) return rng.nextInt(250, 1200);
    if (r < 0.97) return 1000L + 1000L * rng.nextInt(0, 5);
    return 10_000L + 5_000L * rng.nextInt(0, 2);
  }

  private static int sampleLevel(ThreadLocalRandom rng) {
    int r = rng.nextInt(100);
    for (int i = 0; i < LEVEL_CDF.length; i++) if (r < LEVEL_CDF[i]) return i + 1;
    return 5;
  }

  private static double round3(double v) {
    return Math.round(v * 1000.0) / 1000.0;
  }

  private static double clamp(double v, double lo, double hi) {
    return Math.max(lo, Math.min(hi, v));
  }

  private static List<String> buildDealerPool() {
    String[] prefixes = {"NORTH", "OMNI", "KAPI", "ZENI", "AXIS", "HALO", "VERT", "QORE"};
    String[] countries = {"US", "UK"};
    java.util.ArrayList<String> out = new java.util.ArrayList<>(prefixes.length * 5 * countries.length);
    for (String p : prefixes) {
      for (int i = 1; i <= 5; i++) {
        for (String c : countries) {
          out.add(p + String.format(Locale.ROOT, "%02d", i) + "_" + c);
        }
      }
    }
    return List.copyOf(out);
  }

  static int parseInt(String[] args, String flag, int def) {
    for (int i = 0; i + 1 < args.length; i++) {
      if (flag.equals(args[i])) {
        try {
          return Integer.parseInt(args[i + 1]);
        } catch (NumberFormatException ignore) {
          // keep default
        }
      }
    }
    return def;
  }

  /** Tiny shim to avoid bringing in java.util.concurrent.locks just for one parkNanos. */
  static final class LockSupport {
    static void parkNanos(long n) {
      java.util.concurrent.locks.LockSupport.parkNanos(n);
    }
  }
}

File: examples-bin/markets/bond_inventory_producer.py
#!/usr/bin/env python3
"""Anonymised dealer-inventory bid/offer producer for the bond markets example.

Mirrors the field shapes of the user's mrkaxis ``kafka_inventory_producer.py`` — but with
**anonymised** firm codes / market segments and no Aiven SASL config. Writes JSON rows to
``fnd-inventory`` on the local broker brought up by ``docker-compose-kafka.yml``.

Fields:
    companyShortName, instrumentId, side, price, size, spread, level, tier,
    marketSegment, productCD, quoteType, action, processedTs

Run with:
    pip install kafka-python numpy
    python examples-bin/markets/bond_inventory_producer.py
"""

from __future__ import annotations

import argparse
import time
from datetime import datetime, timezone

import numpy as np
from kafka import KafkaProducer  # type: ignore

import kafka_config

SEED = 42
rng = np.random.default_rng(SEED)

# --- Anonymised pools (different from the source pipeline) -------------------
# Firm codes (replacing JPMG / CITI / BOFA / MSCO / UBSW etc.)
DEALER_POOL = np.array(
    [f"{p}{i:02d}_{cc}" for p in ("NORTH", "OMNI", "KAPI", "ZENI", "AXIS", "HALO", "VERT", "QORE")
     for i in range(1, 6) for cc in ("US", "UK")]
)
SIDES = np.array(["BID", "OFFER"])
ACTIONS = np.array(["UPDATE", "DELETE"])
QUOTE_TYPES = np.array(["I", "F"])
TIERS = np.array([1, 2, 3, 4])
MARKET_SEGMENTS = np.array(["IG", "AA", "HY"])   # investment-grade / AA-tier / high-yield
PRODUCT_CDS = np.array(["FNDIG", "FNDAA", "FNDSC", "FNDHS", "FNDHY", "FNDPG"])

LEVEL_PROBS = np.array([0.15, 0.20, 0.35, 0.20, 0.10])
LEVELS = np.array([1, 2, 3, 4, 5])

# Price anchors per segment (mid, std).
SEG_ANCHOR = {"IG": (110.0, 8.0), "AA": (112.0, 7.0), "HY": (92.0, 10.0)}
SEG_BOUNDS = {"IG": (85.0, 140.0), "AA": (90.0, 140.0), "HY": (60.0, 120.0)}


def gen_sizes(n: int) -> np.ndarray:
    bucket = rng.choice(
        ["tiny", "small", "med", "big", "huge"], size=n, p=[0.25, 0.35, 0.25, 0.12, 0.03])
    out = np.empty(n, dtype=np.int64)
    out[bucket == "tiny"] = rng.integers(1, 20, size=(bucket == "tiny").sum())
    out[bucket == "small"] = rng.integers(20, 250, size=(bucket == "small").sum())
    out[bucket == "med"] = rng.integers(250, 1200, size=(bucket == "med").sum())
    out[bucket == "big"] = rng.choice([1000, 2000, 5000], size=(bucket == "big").sum())
    out[bucket == "huge"] = rng.choice([10000, 15000], size=(bucket == "huge").sum())
    return out


def gen_prices(n: int, segs: np.ndarray, sides: np.ndarray) -> np.ndarray:
    out = np.zeros(n, dtype=float)
    for seg in np.unique(segs):
        idx = np.where(segs == seg)[0]
        mean, std = SEG_ANCHOR[seg]
        nudge = np.where(sides[idx] == "BID", -0.6, 0.6)
        out[idx] = rng.normal(mean + nudge, std)
    lo = np.vectorize(lambda s: SEG_BOUNDS[s][0])(segs)
    hi = np.vectorize(lambda s: SEG_BOUNDS[s][1])(segs)
    return np.round(np.clip(out, lo, hi), 3)


def gen_batch(batch: int) -> list[dict]:
    now = datetime.now(timezone.utc)
    iso = now.isoformat().replace("+00:00", "Z")
    companies = rng.choice(DEALER_POOL, size=batch)
    sides = rng.choice(SIDES, size=batch)
    segs = rng.choice(MARKET_SEGMENTS, size=batch)
    instruments = rng.integers(10_000_000, 40_000_000, size=batch, dtype=np.int64)
    prices = gen_prices(batch, segs, sides)
    sizes = gen_sizes(batch)
    spreads = np.where(rng.random(batch) < 0.15, rng.normal(120, 30, size=batch), 0.0).clip(min=0)
    levels = rng.choice(LEVELS, size=batch, p=LEVEL_PROBS)
    tiers = rng.choice(TIERS, size=batch)
    products = rng.choice(PRODUCT_CDS, size=batch)
    quote_types = rng.choice(QUOTE_TYPES, size=batch)
    actions = rng.choice(ACTIONS, size=batch, p=[0.97, 0.03])

    rows = []
    for i in range(batch):
        rows.append(
            dict(
                companyShortName=str(companies[i]),
                instrumentId=int(instruments[i]),
                side=str(sides[i]),
                price=float(prices[i]),
                size=int(sizes[i]),
                spread=float(spreads[i]),
                level=int(levels[i]),
                tier=int(tiers[i]),
                marketSegment=str(segs[i]),
                productCD=str(products[i]),
                quoteType=str(quote_types[i]),
                action=str(actions[i]),
                processedTs=iso,
            )
        )
    return rows


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--rate", type=int, default=500, help="rows per second")
    ap.add_argument("--batch", type=int, default=50, help="rows per Kafka send batch")
    args = ap.parse_args()

    producer = KafkaProducer(**kafka_config.producer_kwargs())
    topic = kafka_config.TOPIC_BOND_INVENTORY
    interval = args.batch / max(args.rate, 1)
    sent = 0
    started = time.time()
    print(f"producing {args.rate} rows/s to '{topic}' (batch={args.batch})")
    try:
        while True:
            for row in gen_batch(args.batch):
                producer.send(topic, value=row)
            sent += args.batch
            if sent % (args.batch * 20) == 0:
                rate = sent / max(time.time() - started, 1e-6)
                print(f"  sent {sent} rows; effective {rate:.0f}/s")
            time.sleep(interval)
    except KeyboardInterrupt:
        print("\nstopping")
    finally:
        producer.flush()
        producer.close()


if __name__ == "__main__":
    main()
